File: adapters/recall-mcp/src/korg_recall_mcp/text.py
"""Turn a korg ledger event into the text string we'll embed for recall.

The shape we expect (from `korg-ingest-claude` and any other adapter that
writes via `make_jsonl_emit`):

    {
      "seq": 42,
      "source_agent": "agent:claude-code#abc",
      "tool_name": "user_prompt" | "llm_inference" | <tool name>,
      "args": {...},
      "result": {...},
      "success": true,
      "duration_ms": 0,
      "triggered_by": 41
    }

The flattening rules below favor signal over noise — we want each
event's text to capture *what it was actually about* in a sentence or
two, so cosine similarity does the heavy lifting on small vectors.
"""

from __future__ import annotations

from typing import Any


_MAX_FIELD_CHARS = 600  # cap any individual field's contribution


def _trim(s: str, n: int = _MAX_FIELD_CHARS) -> str:
    if len(s) <= n:
        return s
    return s[: n - 1] + "…"


def text_for_event(event: dict[str, Any]) -> str:
    """Return a single string describing the event for embedding.

    Returns an empty string if nothing meaningful can be extracted — the
    indexer skips those rather than embedding noise.
    """
    tool = event.get("tool_name", "")
    args = event.get("args") or {}
    result = event.get("result") or {}

    if tool == "user_prompt":
        prompt = args.get("prompt") or args.get("text") or ""
        return _trim(str(prompt))

    if tool == "llm_inference":
        return _trim(str(result.get("text") or ""))

    # Generic tool call: tool_name + key args + a short result preview.
    parts: list[str] = [tool] if tool else []

    # A small set of args usually captures intent; the rest is noise.
    for key in ("file_path", "command", "description", "query", "pattern", "url"):
        v = args.get(key)
        if v:
            parts.append(f"{key}={v}")

    # If no recognized args, fall back to a short JSON-ish dump.
    if len(parts) <= 1 and args:
        try:
            import json as _json

            dumped = _json.dumps(args, default=str)[:300]
            parts.append(dumped)
        except Exception:
            pass

    # Include a short snippet of the result output so e.g. a Read of a file
    # that returned "TODO: rate limiter" is findable by querying "rate limiter".
    output = result.get("output") or result.get("text") or ""
    if isinstance(output, str) and output.strip():
        parts.append(f"→ {output[:300]}")

    text = " ".join(p for p in parts if p)
    return _trim(text)

File: adapters/recall-mcp-ts/src/text.ts
// Turn a korg ledger event into the text string we embed/search.
//
// Mirrors the Python reference implementation at
// adapters/recall-mcp/src/korg_recall_mcp/text.py — same per-event-type
// rules so the two implementations rank the same ledger identically.

export interface LedgerEvent {
  seq?: number;
  source_agent?: string;
  tool_name?: string;
  args?: Record<string, unknown> | null;
  result?: Record<string, unknown> | null;
  success?: boolean;
  duration_ms?: number;
  triggered_by?: number | null;
}

const MAX_FIELD_CHARS = 600;

function trim(s: string, n: number = MAX_FIELD_CHARS): string {
  if (s.length <= n) return s;
  return s.slice(0, n - 1) + "…";
}

function asString(v: unknown): string {
  if (typeof v === "string") return v;
  if (v == null) return "";
  return String(v);
}

/**
 * Return a single string describing the event for embedding / substring search.
 * Empty string when nothing meaningful can be extracted — the indexer
 * skips those rather than embedding noise.
 */
export function textForEvent(event: LedgerEvent): string {
  const tool = event.tool_name ?? "";
  const args = (event.args ?? {}) as Record<string, unknown>;
  const result = (event.result ?? {}) as Record<string, unknown>;

  if (tool === "user_prompt") {
    const prompt = asString(args["prompt"] ?? args["text"] ?? "");
    return trim(prompt);
  }

  if (tool === "llm_inference") {
    return trim(asString(result["text"] ?? ""));
  }

  // Generic tool call: tool_name + a few key args + a result preview.
  const parts: string[] = [];
  if (tool) parts.push(tool);

  for (const key of ["file_path", "command", "description", "query", "pattern", "url"]) {
    const v = args[key];
    if (v != null && v !== "") {
      parts.push(`${key}=${asString(v)}`);
    }
  }

  // Fallback for unknown args: a short JSON dump.
  if (parts.length <= 1 && Object.keys(args).length > 0) {
    try {
      parts.push(JSON.stringify(args).slice(0, 300));
    } catch {
      // ignore
    }
  }

  // Result snippet so e.g. a Read of a file that returned "TODO: rate limiter"
  // is findable by querying "rate limiter".
  const output = result["output"] ?? result["text"] ?? "";
  const outStr = asString(output);
  if (outStr.trim().length > 0) {
    parts.push(`→ ${outStr.slice(0, 300)}`);
  }

  const text = parts.filter((p) => p.length > 0).join(" ");
  return trim(text);
}
